drivers/photo-electronic/adpd410x/adpd410x.c

```diff
@@ -0,0 +1,659 @@
+	}
+}
+
+/**
+ * @brief Do a read and write of a register to update only part of a register.
+ * @param dev - Device handler.
+ * @param address - Address of the register.
+ * @param data - Value to be written to the device.
+ * @param mask - Mask of the bit field to update.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_reg_write_mask(struct adpd410x_dev *dev, uint16_t address,
+				uint16_t data, uint16_t mask)
+{
+	int32_t ret;
+	uint16_t reg_val;
+	uint32_t bit_pos;
+
+	ret = adpd410x_reg_read(dev, address, &reg_val);
+	if (ret != SUCCESS)
+		return FAILURE;
+	reg_val &= ~mask;
+	bit_pos = find_first_set_bit((uint32_t)mask);
+	reg_val |= (data << bit_pos) & mask;
+
+	return adpd410x_reg_write(dev, address, reg_val);
+}
+
+/**
+ * @brief Update the clocking option of the device.
+ * @param dev - Device handler.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+static int32_t adpd410x_get_clk_opt(struct adpd410x_dev *dev)
+{
+	int32_t ret;
+	uint16_t reg_temp;
+
+	ret = adpd410x_reg_read(dev, ADPD410X_REG_SYS_CTL, &reg_temp);
+	if(ret != SUCCESS)
+		return FAILURE;
+	dev->clk_opt = (reg_temp & BITM_SYS_CTL_ALT_CLOCKS) >>
+		       BITP_SYS_CTL_ALT_CLOCKS;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Do a software reset.
+ * @param dev - Device handler.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_reset(struct adpd410x_dev *dev)
+{
+	int32_t ret;
+
+	ret = adpd410x_reg_write_mask(dev, ADPD410X_REG_SYS_CTL, 1,
+				      BITM_SYS_CTL_SW_RESET);
+	if(ret != SUCCESS)
+		return ret;
+
+	return adpd410x_get_clk_opt(dev);
+}
+
+/**
+ * @brief Set operation mode.
+ * @param dev - Device handler.
+ * @param mode - New operation mode.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_set_opmode(struct adpd410x_dev *dev,
+			    enum adpd410x_opmode mode)
+{
+	return adpd410x_reg_write_mask(dev, ADPD410X_REG_OPMODE, mode,
+				       BITM_OPMODE_OP_MODE);
+}
+
+/**
+ * @brief Set number of active time slots.
+ * @param dev - Device handler.
+ * @param timeslot_no - Last time slot to be enabled.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_set_last_timeslot(struct adpd410x_dev *dev,
+				   enum adpd410x_timeslots timeslot_no)
+{
+	return adpd410x_reg_write_mask(dev, ADPD410X_REG_OPMODE,
+				       timeslot_no, BITM_OPMODE_TIMESLOT_EN);
+}
+
+/**
+ * @brief Set device sampling frequency.
+ * @param dev - Device handler.
+ * @param sampling_freq - New sampling frequency.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_set_sampling_freq(struct adpd410x_dev *dev,
+				   uint32_t sampling_freq)
+{
+	int32_t ret;
+	uint32_t reg_load;
+	uint16_t reg_temp;
+
+	if((dev->clk_opt == ADPD410X_INTLFO_INTHFO) ||
+	    (dev->clk_opt == ADPD410X_INTLFO_EXTHFO)) {
+		ret = adpd410x_reg_read(dev, ADPD410X_REG_SYS_CTL, &reg_temp);
+		if(ret != SUCCESS)
+			return FAILURE;
```

Comment:
I think we should use <errno.h> defines in all new drivers. (<-- applies throughout all the file)
In this case though it should be `return ret`.